pilot/pkg/networking/grpcgen/cds.go

```diff
@@ -0,0 +1,270 @@
+	trafficPolicy := v1alpha3.MergeTrafficPolicy(nil, destinationRule.TrafficPolicy, b.port)
+
+	// setup default cluster
+	b.applyPolicy(defaultCluster, trafficPolicy)
+
+	// subset clusters
+	if len(destinationRule.Subsets) > 0 {
+		subsetClusters = make([]*clusterv3.Cluster, 0, len(destinationRule.Subsets))
+		for _, subset := range destinationRule.Subsets {
+			subsetKey := subsetClusterKey(subset.Name, string(b.hostname), b.portNum)
+			if !b.filter.Contains(subsetKey) {
+				continue
+			}
+			c := edsCluster(subsetKey)
+			trafficPolicy := v1alpha3.MergeTrafficPolicy(trafficPolicy, subset.TrafficPolicy, b.port)
+			b.applyPolicy(c, trafficPolicy)
+			subsetClusters = append(subsetClusters, c)
+		}
+	}
+
+	return
+}
+
+// applyPolicy mutates the give cluster (if not-nil) so that the given merged traffic policy applies.
+func (b *clusterBuilder) applyPolicy(c *clusterv3.Cluster, trafficPolicy *networking.TrafficPolicy) {
+	// cluster can be nil if it wasn't requested
+	if c == nil || trafficPolicy == nil {
+		return
+	}
+	b.applyTLS(c, trafficPolicy)
+	b.applyLoadBalancing(c, trafficPolicy)
+	// TODO status or log when unsupported features are included
+}
+
+func (b *clusterBuilder) applyLoadBalancing(c *clusterv3.Cluster, policy *networking.TrafficPolicy) {
+	switch policy.LoadBalancer.GetSimple() {
+	case networking.LoadBalancerSettings_ROUND_ROBIN:
+	// ok
+	default:
+		log.Warnf("cannot apply LbPolicy %s to %s", policy.LoadBalancer.GetSimple(), b.node.ID)
+	}
+
+	// TODO https://github.com/grpc/proposal/blob/master/A42-xds-ring-hash-lb-policy.md
+}
+
+func (b *clusterBuilder) applyTLS(c *clusterv3.Cluster, policy *networking.TrafficPolicy) {
+	// TODO check for automtls
+	mode := networking.ClientTLSSettings_ISTIO_MUTUAL
+	if settings := policy.GetTls(); settings != nil {
+		mode = settings.GetMode()
+	}
+
+	switch mode {
+	case networking.ClientTLSSettings_DISABLE:
+		// nothing to do
+	case networking.ClientTLSSettings_SIMPLE:
+		// TODO support this
+	case networking.ClientTLSSettings_MUTUAL:
+		// TODO support this
+	case networking.ClientTLSSettings_ISTIO_MUTUAL:
+		tlsCtx := buildTLSContext(b.push.ServiceAccounts[b.hostname][b.portNum])
+		c.TransportSocket = &envoy_config_core_v3.TransportSocket{
+			Name:       transportSocketName,
+			ConfigType: &envoy_config_core_v3.TransportSocket_TypedConfig{TypedConfig: util.MessageToAny(tlsCtx)},
+		}
+	}
+
+}
+
+// TransportSocket proto message has a `name` field which is expected to be set to exactly this value by the
+// management server (see grpc/xds/internal/client/xds.go securityConfigFromCluster).
+const transportSocketName = "envoy.transport_sockets.tls"
+
+// buildTLSContext creates a TLS context that assumes 'default' name, and credentials/tls/certprovider/pemfile
+// (see grpc/xds/internal/client/xds.go securityConfigFromCluster).
+func buildTLSContext(sans []string) *tlsv3.UpstreamTlsContext {
```

Comment:
seems generic enough that maybe name should be passed in instead of set to a default value

Reply:
The `InstanceName`? This is tightly coupled for now with the CertProvider instance that we set up here:

https://github.com/istio/istio/blob/876407b38ce3fa8221d44de66f16a15d38fba799/pkg/istio-agent/grpcxds/grpc_bootstrap.go#L138


I can add some comments to make this clear.

Reply:
yeah maybe thats a good idea